src/hotspot/share/jfr/leakprofiler/chains/dfsClosure.cpp

```diff
@@ -60,52 +62,65 @@ void DFSClosure::find_leaks_from_root_set(EdgeStore* edge_store,
   dfs._max_depth = 1;
   RootSetClosure<DFSClosure> rs(&dfs);
   rs.process();
+  dfs.drain_probe_stack();
 
   // Depth-first search
   dfs._max_depth = max_dfs_depth;
   dfs._ignore_root_set = true;
   rs.process();
+  dfs.drain_probe_stack();
 }
 
 DFSClosure::DFSClosure(EdgeStore* edge_store, JFRBitSet* mark_bits, const Edge* start_edge)
   :_edge_store(edge_store), _mark_bits(mark_bits), _start_edge(start_edge),
   _max_depth(max_dfs_depth), _depth(0), _ignore_root_set(false) {
 }
 
-void DFSClosure::closure_impl(UnifiedOopRef reference, const oop pointee) {
-  assert(pointee != nullptr, "invariant");
-  assert(!reference.is_null(), "invariant");
+void DFSClosure::drain_probe_stack() {
 
-  if (GranularTimer::is_finished()) {
-    return;
-  }
+  while (!_probe_stack.is_empty()) {
 
-  if (_depth == 0 && _ignore_root_set) {
-    // Root set is already marked, but we want
-    // to continue, so skip is_marked check.
-    assert(_mark_bits->is_marked(pointee), "invariant");
-    _reference_stack[_depth] = reference;
-  } else {
-    if (_mark_bits->is_marked(pointee)) {
+    if (GranularTimer::is_finished()) {
       return;
     }
-    _mark_bits->mark_obj(pointee);
-    _reference_stack[_depth] = reference;
-    // is the pointee a sample object?
-    if (pointee->mark().is_marked()) {
-      add_chain();
+
+    const ProbeStackItem psi = _probe_stack.pop();
+
+    const UnifiedOopRef reference = psi.r;
+    assert(!reference.is_null(), "invariant");
+    const oop pointee = reference.dereference();
+    assert(pointee != nullptr, "invariant");
+
+    _depth = psi.d;
+
+    if (_depth == 0 && _ignore_root_set) {
+      // Root set is already marked, but we want
+      // to continue, so skip is_marked check.
+      assert(_mark_bits->is_marked(pointee), "invariant");
+      _reference_stack[_depth] = reference;
+    } else {
+      if (_mark_bits->is_marked(pointee)) {
+        return;
```

Comment:
I think this improvement is a good idea!  But maybe this line should be replaced with a `continue`, otherwise we can terminate the DFS prematurely and skip evaluation of other chains extending from other references already pushed to the stack.

Reply:
Hmm, Erik also commented on this. The obviously right answer is that we need to continue here.

Looking at `TestJcmdDumpPathToGCRoots.java`, it is a bit simple in that it just builds up a single long list of object. We will have saturated the max_dfs_depth long before the end of the chain is reached, and its a single long chain which may not hit the return-bug, and the test does not seem to count the expected number of events. 

I'll check if that can be improved.